### PR TITLE
order by implementation

### PR DESCRIPTION
**Pull request by onurkuruu**
*Issue #47 *

* Support for `NULLS FIRST`, `NULLS LAST` keywords
* Support for ordering `null/missing`, `bool`, `number`, `date/timestamp/time`, `string` types
* A bug fixed which `SqlParser` cannot parse if the statement contains `ORDER BY` which comes after `GROUP BY`
* Support for ordering the results;
	* Normal results
	* Aggregated results without grouping (GROUP BY)
	* Grouped results

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt [hunk 1]
```diff
@@ -1982,6 +2002,35 @@ internal class EvaluatingCompiler(
...
+                is PartiqlAst.OrderingSpec.Desc -> OrderingType.DESC
+            }
+
+            val nullsSpec = when (it.nulls) {
+                is PartiqlAst.NullsSpec.NullsFirst -> NullsType.FIRST
+                is PartiqlAst.NullsSpec.NullsLast -> NullsType.LAST
+            }
+
+            CompiledOrderByItem(orderSpec, nullsSpec, compileAstExpr(it.expr))
+        }
```
Comment (alancai98) [inline lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt:2032]: I think [NaturalExprValueComparators](https://github.com/partiql/partiql-lang-kotlin/blob/67e3adf452d1d411aaba71de2635e12bb25bca2e/lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt) can be used here with some slight modification to case on the `OrderingSpec`. Using that could remove the need for `CompareExtentions.kt` and provide the implementation for some of the type comparisons.

Also, `NaturalExprValueComparators` has a lot of tests in [NaturalExprValueComparatorsTest.kt](https://github.com/partiql/partiql-lang-kotlin/blob/67e3adf452d1d411aaba71de2635e12bb25bca2e/lang/test/org/partiql/lang/eval/NaturalExprValueComparatorsTest.kt), which can reduce the amount of tests needed to be written.

File: lang/test/org/partiql/lang/eval/EvaluatingCompilerOrderByTests.kt [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+            // should order by supplierId_nulls nulls last
+            EvaluatorTestCase(
+                "SELECT productId FROM products_sparse ORDER BY supplierId_nulls NULLS LAST, productId",
+                "<<{'productId': 1}, {'productId': 2}, {'productId': 3}, {'productId': 6}, {'productId': 7}, {'productId': 4}, {'productId': 5}, {'productId': 8}, {'productId': 9}, {'productId': 10}>>"
+            ),
+            // should order by supplierId_nulls nulls first
+            EvaluatorTestCase(
+                "SELECT productId FROM products_sparse ORDER BY supplierId_nulls NULLS FIRST, productId",
+                "<<{'productId': 4}, {'productId': 5}, {'productId': 8}, {'productId': 9}, {'productId': 10}, {'productId': 1}, {'productId': 2}, {'productId': 3}, {'productId': 6}, {'productId': 7}>>"
+            )
```
Comment (alancai98) [inline lang/test/org/partiql/lang/eval/EvaluatingCompilerOrderByTests.kt:88]: nit: may be helpful to include `supplierId_nulls` in the select projection to verify the results

**Review by alancai98 — CHANGES_REQUESTED** [review]
Thanks for the contribution and putting together this PR! This is definitely a great start to the implementation! I noticed a few things that need addressed:
- changing default behavior when ordering spec (i.e. `ASC`/`DESC`) is not specified
- using `NaturalExprValueComparators` to perform the `ORDER BY` comparison between the different types
- some more tests (parsing, evaluation)
Other than that, only had some nits/typos.

**Comment by onurkuruu** [comment]
Hi @alancai98, I have pushed a commit to the branch. I tried to use `NaturalExprValueComparators` to order the rows but I am not sure about the code changes if this is what you wanted. If this is the correct approach you wanted, I can add more test cases to `NaturalExprValueComparatorsTest` to cover new changes.

Here is the what the commit contains;

* fixed some typo
* added `asc/desc` support to `NaturalExprValueComparators`
* using `NaturalExprValueComparators` when ordering rows
* `getQueryThunk` returning list if order by specified otherwise bag
* added more test cases to `SqlParserTest` and `EvaluatingCompilerOrderByTests`

File: lang/src/org/partiql/lang/ast/ast.kt [hunk 3]
```diff
@@ -1055,6 +1056,13 @@ enum class OrderingSpec {
     DESC
 }
 
+/** Nulls specification */
+enum class NullsSpec {
+    /** Represents */
```
[the remark this answers is not in the document]
Comment (onurkuruu) [reply lang/src/org/partiql/lang/ast/ast.kt]: I have added this with new commit.

File: lang/src/org/partiql/lang/syntax/SqlParser.kt [hunk 4]
```diff
@@ -132,6 +132,7 @@ class SqlParser(
         ORDER_BY,
         SORT_SPEC,
         ORDERING_SPEC,
+        NULLS,
```
[the remark this answers is not in the document]
Comment (onurkuruu) [reply lang/src/org/partiql/lang/syntax/SqlParser.kt]: I have added this with new commit.

File: lang/test/org/partiql/lang/errors/ParserErrorsTest.kt [hunk 5]
```diff
@@ -1134,6 +1134,90 @@ class ParserErrorsTest : SqlParserTestBase() {
         )
     }
 
+    @Test
```
[the remark this answers is not in the document]
Comment (onurkuruu) [reply lang/test/org/partiql/lang/errors/ParserErrorsTest.kt:1137]: Thanks!

File: lang/test/org/partiql/lang/syntax/SqlParserTest.kt [hunk 6]
```diff
@@ -1696,7 +1696,8 @@ class SqlParserTest : SqlParserTestBase() {
                 (order_by 
                     (sort_spec 
                         (id rk1 (case_insensitive) (unqualified)) 
-                        (asc)))))
+                        (asc)
```
[the remark this answers is not in the document]
Comment (onurkuruu) [reply lang/test/org/partiql/lang/syntax/SqlParserTest.kt:1699]: I have some more test cases which also covers `NULLS FIRST`/`NULLS LAST`

File: lang/src/org/partiql/lang/syntax/SqlParser.kt [hunk 7]
```diff
@@ -577,8 +578,13 @@ class SqlParser(
...
                                 when (it.children.size) {
-                                    1 -> sortSpec(it.children[0].toAstExpr(), asc())
-                                    2 -> sortSpec(it.children[0].toAstExpr(), it.children[1].toOrderingSpec())
+                                    1 -> sortSpec(it.children[0].toAstExpr(), asc(), nullsLast())
+                                    2 -> when (it.children[1].type) {
+                                        ParseType.ORDERING_SPEC -> sortSpec(it.children[0].toAstExpr(), it.children[1].toOrderingSpec(), nullsLast())
+                                        ParseType.NULLS -> sortSpec(it.children[0].toAstExpr(), asc(), it.children[1].toNullsSpec())
+                                        else -> errMalformedParseTree("Invalid ordering expressions syntax")
+                                    }
+                                    3 -> sortSpec(it.children[0].toAstExpr(), it.children[1].toOrderingSpec(), it.children[2].toNullsSpec())
```
[the remark this answers is not in the document]
Comment (onurkuruu) [reply lang/src/org/partiql/lang/syntax/SqlParser.kt]: I have updated here as you defined above

File: lang/src/org/partiql/lang/errors/ErrorCode.kt [hunk 8]
```diff
@@ -557,6 +557,22 @@ enum class ErrorCode(
...
 
+    EVALUATOR_ORDER_BY_NULL_COMPARATOR(
+        ErrorCategory.EVALUATOR,
+        LOCATION,
+        ""
+    ) {
+        override fun getErrorMessage(errorContext: PropertyValueMap?): String = ""
+    },
+
+    EVALUATOR_ORDER_BY_ENVIRONMENT_CANNOT_RESOLVED(
```
[the remark this answers is not in the document]
Comment (onurkuruu) [reply lang/src/org/partiql/lang/errors/ErrorCode.kt]: I have changed here with `EVALUATOR_ENVIRONMENT_CANNOT_BE_RESOLVED` since I have changed the logic of the code which use this error message.

File: lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt [hunk 9]
```diff
@@ -55,9 +55,17 @@ import org.partiql.lang.util.isZero
...
-enum class NaturalExprValueComparators(private val nullOrder: NullOrder) : Comparator<ExprValue> {
-    NULLS_FIRST(NullOrder.FIRST),
-    NULLS_LAST(NullOrder.LAST);
+enum class NaturalExprValueComparators(private val order: Order, private val nullOrder: NullOrder) : Comparator<ExprValue> {
+    NULLS_FIRST_ASC(Order.ASC, NullOrder.FIRST),
+    NULLS_FIRST_DESC(Order.DESC, NullOrder.FIRST),
+    NULLS_LAST_ASC(Order.ASC, NullOrder.LAST),
+    NULLS_LAST_DESC(Order.DESC, NullOrder.LAST);
+
+    /** Whether or not null values come first or last. */
```
Comment (alancai98) [inline lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt:64]: typo: doc comment should talk about the ordering spec not null order spec

File: lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt [hunk 10]
```diff
@@ -55,9 +55,17 @@ import org.partiql.lang.util.isZero
  *  * `BAG` values come finally (except with [NullOrder.NULLS_LAST]), and their values
  *    compare lexicographically based on the *sorted* child elements.
  */
-enum class NaturalExprValueComparators(private val nullOrder: NullOrder) : Comparator<ExprValue> {
-    NULLS_FIRST(NullOrder.FIRST),
-    NULLS_LAST(NullOrder.LAST);
+enum class NaturalExprValueComparators(private val order: Order, private val nullOrder: NullOrder) : Comparator<ExprValue> {
```
Comment (alancai98) [inline lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt:58]: nit (documentation): since we're adding another argument to the public constructor, could add a kdoc comment for the new parameter, `order`

File: lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt [hunk 11]
```diff
@@ -1998,6 +2025,92 @@ internal class EvaluatingCompiler(
...
+            it.orderingSpec
+                ?: errNoContext(
+                    "SortSpec.orderingSpec was not specified",
+                    errorCode = ErrorCode.INTERNAL_ERROR,
+                    internal = true
+                )
+
+            it.nullsSpec
+                ?: errNoContext(
+                    "SortSpec.nulls was not specified",
```
Comment (alancai98) [inline lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt:2039]: typo: should say `SortSpec.nullsSpec`

```suggestion
                    "SortSpec.nullsSpec was not specified",
```

File: lang/test/org/partiql/lang/eval/EvaluatingCompilerOrderByTests.kt [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+                "SELECT sellerId FROM orders GROUP BY sellerId ORDER BY sellerId ASC",
+                "[{'sellerId': 1}, {'sellerId': 2}]"
+            ),
+            // should group and order by desc sellerId
+            EvaluatorTestCase(
+                "SELECT sellerId FROM orders GROUP BY sellerId ORDER BY sellerId DESC",
+                "[{'sellerId': 2}, {'sellerId': 1}]"
+            )
+        )
+    }
```
Comment (alancai98) [inline lang/test/org/partiql/lang/eval/EvaluatingCompilerOrderByTests.kt:100]: This is a good start for the evaluation tests. I'll discuss further with the team about what level of coverage seems reasonable as `NaturalExprValueComparatorsTest` covers a lot of edge cases. Can consider adding more tests with
- `MISSING`, `NULL` and `MISSING`
- different data types (not just numbers)

**Review by alancai98 — COMMENTED** [review]
Changes since the initial commit look good. Identified a few nits/typos and there could be some more evaluation tests with other data types.

I think the approach using `NaturalExprValueComparators` looks reasonable. What other tests were you considering adding?

**Comment by onurkuruu** [comment]
> Changes since the initial commit look good. Identified a few nits/typos and there could be some more evaluation tests with other data types.
> 
> I think the approach using `NaturalExprValueComparators` looks reasonable. What other tests were you considering adding?

I thought , I could add some tests to cover `ASC/DESC` feature of `NaturalExprValueComparators` to see if everything working as expected. Actually, EvaluatingCompilerOrderByTests covers `ASC/DESC` feature of `NaturalExprValueComparators` but I thought, It would be good also adding some test cases for `NULLS_FIRST_DESC` and `NULLS_LAST_DESC`.

**Comment by alancai98** [comment]
>I thought , I could add some tests to cover `ASC/DESC` feature of `NaturalExprValueComparators` to see if everything working as expected. Actually, EvaluatingCompilerOrderByTests covers `ASC/DESC` feature of `NaturalExprValueComparators` but I thought, It would be good also adding some test cases for `NULLS_FIRST_DESC` and `NULLS_LAST_DESC`. 

Yeah I think those would be good to add.

---

Regarding the tests for [EvaluatingCompilerOrderByTests.kt](https://github.com/partiql/partiql-lang-kotlin/blob/2b2a1f26951bee018af377c91d86832a65a635ba/lang/test/org/partiql/lang/eval/EvaluatingCompilerOrderByTests.kt), we think more tests should be added for the end-to-end evaluator behavior (even though there are unit tests for [NaturalExprValueComparators](https://github.com/partiql/partiql-lang-kotlin/blob/2b2a1f26951bee018af377c91d86832a65a635ba/lang/test/org/partiql/lang/eval/NaturalExprValueComparatorsTest.kt)). Tests should ensure the ORDER BY ordering follows what's specified in [section 12.2](https://partiql.org/assets/PartiQL-Specification.pdf#page=44) of the spec. As a summary, should provide the following tests:
- tests between different data types
  - assuming `ASC` and `NULLS FIRST`, check end-to-end evaluation follows the order: unknowns (null, missing), bool, numbers, date, time, timestamp, text, LOB types, lists, structs, bags)
- edge case tests for values of the same data type
  - `false` before `true`
  - `nan` before `-inf`, then numeric values, then `+inf`
  - `timestamp`s compared by absolute point of time
  - `text` types compared by lexicographical ordering of Unicode scalar
  - `LOB` types follow their lexicographical ordering by octet
  - `list`s compared lexicographically based on comparison of elements, recursively
    - also, shorter arrays come first
  - `struct`s compared lexicographically based on sorted attributes (first by key then by value)
  - `bag`s compared as sorted lists

Feel free to use the data provided in the tests from [NaturalExprValueComparatorsTest.k](https://github.com/partiql/partiql-lang-kotlin/blob/2b2a1f26951bee018af377c91d86832a65a635ba/lang/test/org/partiql/lang/eval/NaturalExprValueComparatorsTest.kt#L28-L291). If it's any easier, you can also include the data directly in the from source rather than the evaluation session. E.g.:

```
SELECT * FROM [{ 'a': false }, { 'a': true }] ORDER BY a
```

File: lang/resources/org/partiql/type-domains/partiql.ion [hunk 13]
```diff
@@ -217,14 +217,20 @@
         (product order_by sort_specs::(* sort_spec 1))
 
         // <expr> [ASC | DESC] ?
```
[the remark this answers is not in the document]
Comment (onurkuruu) [reply lang/resources/org/partiql/type-domains/partiql.ion]: fixed with new commit

File: lang/test/org/partiql/lang/syntax/SqlParserTest.kt [hunk 14]
```diff
@@ -1777,12 +1782,155 @@ class SqlParserTest : SqlParserTestBase() {
...
+                        (nulls_last))
                     (sort_spec 
                         (id rk2 (case_insensitive) (unqualified)) 
-                        (desc)))))
+                        (desc)
+                        (nulls_first)))))
         """
     )
+
+    @Test
```
[the remark this answers is not in the document]
Comment (onurkuruu) [reply lang/test/org/partiql/lang/syntax/SqlParserTest.kt:1794]: Thanks!

File: lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt (hunk 9 above)
Comment (onurkuruu) [reply lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt]: fixed with new commit

File: lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt (hunk 10 above)
Comment (onurkuruu) [reply lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt:61]: fixed with new commit

File: lang/test/org/partiql/lang/eval/EvaluatingCompilerOrderByTests.kt (hunk 12 above)
Comment (onurkuruu) [reply lang/test/org/partiql/lang/eval/EvaluatingCompilerOrderByTests.kt:273]: I have added more test cases which covers different data types and edge cases

**Comment by onurkuruu** [comment]
Thanks for your review @alancai98 , I have pushed a new commit for your comments.
* Added necessary kdoc comments 
* Fixed some typo
* Added more test cases which covers different data types and edge cases as defined at specifications for `EvaluatingCompilerOrderByTests` 
* Refactored `NaturalExprValueComparatorsTest` to cover desc feature of `NaturalExprValueComparators`

File: lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt [hunk 15]
```diff
@@ -157,6 +170,13 @@ enum class NaturalExprValueComparators(private val nullOrder: NullOrder) : Compa
         }
     }
 
+    private fun <T> orderLeftRight(order: Order, left: T, right: T): Pair<T, T> {
```
Comment (alancai98) [inline lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt:173]: Swapping the left and right side ordering is what we want when comparing between two of the **same** data types for `DESC`. But since we also want values of different types to be sorted in descending order, this approach won't work. To resolve this without repeating too much of the logic, we could break the old version of `compare` into its own function and return the inverse result.

See this gist with one approach to reverse the result of comparison for `DESC`: https://gist.github.com/alancai98/c69fc04f8c4c41177c67c8ffa064012c

**Review by alancai98 — COMMENTED** [review]
@onurkuruu, thanks for adding so many tests! While going through them, I noticed that the semantics of `ORDER BY`'s `DESC` are not entirely correct when comparing between different data types. I tried to explain some of the reasoning and provide a fix in the comments.

Feel free to let me know if anything's confusing or if you have further questions!

File: lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt (hunk 15 above)
Comment (onurkuruu) [reply lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt]: I have committed the code from your gist and changed the tests a bit. Now we are comparing different data types with the following order (Nulls/Missing can be changed by ordering and nulls spec)
* bools, numbers, timestamp, text ... for ascending order
* bag, tuple, array, lob for descending order

File: lang/test/org/partiql/lang/eval/EvaluatingCompilerOrderByTests.kt [hunk 16]
```diff
@@ -0,0 +1,278 @@
...
+            EvaluatorTestCase(
+                """SELECT * FROM [{'a': `{{"Z"}}`}, {'a': `{{"a"}}`}, {'a': `{{"A"}}`}, {'a': `{{"z"}}`}] ORDER BY a""",
+                """[{'a': `{{"A"}}`}, {'a': `{{"Z"}}`}, {'a': `{{"a"}}`}, {'a': `{{"z"}}`}]"""
+            ),
+            // LOB types should ordered (DESC)
+            EvaluatorTestCase(
+                """SELECT * FROM [{'a': `{{"Z"}}`}, {'a': `{{"a"}}`}, {'a': `{{"A"}}`}, {'a': `{{"z"}}`}] ORDER BY a DESC""",
+                """[{'a': `{{"z"}}`}, {'a': `{{"a"}}`}, {'a': `{{"Z"}}`}, {'a': `{{"A"}}`}]"""
+            ),
+            // sorter array comes first (ASC)
```
Comment (alancai98) [inline lang/test/org/partiql/lang/eval/EvaluatingCompilerOrderByTests.kt:212]: typo: shorter
```suggestion
            // shorter array comes first (ASC)
```

File: lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt [hunk 17]
```diff
@@ -104,8 +115,10 @@ enum class NaturalExprValueComparators(private val nullOrder: NullOrder) : Compa
         val rIter = right.iterator()
 
         while (lIter.hasNext() && rIter.hasNext()) {
-            val lChild = lIter.next()
-            val rChild = rIter.next()
+            val (lChild, rChild) = when(order) {
```
Comment (alancai98) [inline lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt:118]: Seems like GitHub actions builds aren't run for forks as mentioned in #561. When I build locally, this line caused a ktlint error. Fixable by running `./gradlew ktlintFormat` or applying the following suggestion:

```suggestion
            val (lChild, rChild) = when (order) {
```

Comment (alancai98) [reply lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt:118]: Similarly with L142.

File: lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt [hunk 18]
```diff
@@ -126,11 +139,15 @@ enum class NaturalExprValueComparators(private val nullOrder: NullOrder) : Compa
     ): Int {
         val pairCmp = object : Comparator<Pair<T, Int>> {
             override fun compare(o1: Pair<T, Int>, o2: Pair<T, Int>): Int {
-                val cmp = entityCmp.compare(o1.first, o2.first)
+                val (leftPair, rightPair) = when(order) {
```
Comment (alancai98) [inline lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt:142]: From the ktlint error:

```suggestion
                val (leftPair, rightPair) = when (order) {
```

File: lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt [hunk 19]
```diff
@@ -294,4 +311,22 @@ enum class NaturalExprValueComparators(private val nullOrder: NullOrder) : Compa
 
         throw IllegalStateException("Could not compare: $left and $right")
     }
+
+    // can think of `DESC` as the converse/result of `ASC`
```
Comment (alancai98) [inline lang/src/org/partiql/lang/eval/NaturalExprValueComparators.kt:315]: Was a typo from my .gist:

```suggestion
    // can think of `DESC` as the converse/reverse of `ASC`
```

**Review by alancai98 — COMMENTED** [review]
Only have a few small nits/style fixes. Should be ready to approve once addressed

**Comment by onurkuruu** [comment]
@alancai98 thanks for your review again. It was pleasure to contribute with your support. What are the steps of releasing new features of  PartiQL? When we can use the PartiQL with order by support with official release?

**Review by alancai98 — APPROVED** [review]

**Comment by alancai98** [comment]
Thanks again for the contribution and the quick responses to feedback!

> What are the steps of releasing new features of PartiQL? When we can use the PartiQL with order by support with official release?

We typically publish new PartiQL versions through [maven](https://search.maven.org/search?q=a:partiql-lang-kotlin). There's currently no release schedule in place, but we can perform the release whenever a customer/user needs the latest version. Do you need this latest version released to maven?

**Comment by onurkuruu** [comment]
Yes, we need. It would be great to have it.

**Comment by alancai98** [comment]
>Yes, we need. It would be great to have it.

Got it. We should be able to start the release process today. I'll followup here w/ any updates.

**Comment by alancai98** [comment]
@onurkuruu the latest release, [v0.6.0](https://github.com/partiql/partiql-lang-kotlin/releases/tag/v0.6.0-alpha), was published to [Maven](https://search.maven.org/artifact/org.partiql/partiql-lang-kotlin/0.6.0/jar) earlier today. Feel free to let us know if you have any issues using it!

**Comment by onurkuruu** [comment]
Thanks @alancai98, I can use `Order By` now. I also want to add alias support for `Order By` and `Group By`. Currently, following queries are not working because of missing bindings.
* `SELECT a as b FROM table ORDER BY b`
* `SELECT a as b FROM table GROUP BY b`

Do you mind if I open a new PR to support for aliases, if can implement? or Do I need to open an issue for this?

**Comment by alancai98** [comment]
Hey @onurkuruu. Sorry have been away due to a bout with COVID. I think it would be good to have select aliases available to `GROUP BY` and `ORDER BY` clauses. I believe PostgreSQL, MySQL, and SQLite support this. I created a GitHub issue #571 from your previous comment.
